scrapy_poet/injection.py

```diff
@@ -141,34 +142,133 @@ def build_plan(self, request: Request) -> andi.Plan:
             overrides=self.overrides_registry.overrides_for(request).get,
         )
 
+    def provider_requirements(self, request: Request, plan: andi.Plan) -> Set[Any]:
+        """Return a set of classes which indicate any requirements needed by a
+        provider in order to successfully provide for the given ``request`` and
+        ``plan``.
+        """
+        provider_requirements = set()
+        for cls, _ in plan.dependencies:
+            for provider in self.providers:
+                if not provider.is_provided(cls):
+                    continue
+                classes = provider.requirements_for(cls, request)
+                if classes:
+                    provider_requirements.update(set(classes))
+        return provider_requirements
+
+    @inlineCallbacks
+    def build_provider_requirements(
+        self,
+        request: Request,
+        response: Response,
+        plan: andi.Plan,
+    ):
+        """This builds out any requirements that a provider might need before
+        calling them.
+
+        The instances that are built here would later be used in andi's
+        'externally_provided' parameter when calling the providers.
+        """
+
+        provider_requirements_instances = {}
+        provider_requirements = self.provider_requirements(request, plan)
+
+        # This recursively gets the provider requirements. For example, a PO
+        # that has an item dependency that needs another PO to produce it.
+        for prov_req in provider_requirements:
+            sub_plan = andi.plan(
+                prov_req,
+                is_injectable=is_injectable,
+                externally_provided=self.is_class_provided_by_any_provider,
+            )
+
+            try:
+                instances, provider_instances = yield from self.build_instances(
+                    request, response, sub_plan
+                )
+            except RecursionError:
```

Comment:
I explored an option by storing the past seen plans and checking it if a deadlock has occurred. However, that pollutes the code by passing a lot of list references to the recursive function calls. It also harder to cover and test a lot of edge cases which deadlocks occur on far away items (e.g. 10 nodes).

I tried adding a context manager to decrease the recursion limit via `sys.setrecursionlimit()` but it could be fragile since we need to balance the current recursion depth with the said limit.

Settling on this solution for now as it's simpler and cleaner. I think we can live with the tradeoffs of having a deeper callstack (max default of 1000) as proactively storing and comparing the past seen plans for deadlock detection also consumes memory and compute time.